File: js-src/src/storage/version-control/Commit.ts
import { Delta, DeltaData } from "../../model/Delta";

export interface CommitMetadataData {
    id: string;
    parent_id: string;
    message: string;
    timestamp: number;
    snapshot_hash: string;
    delta_data?: never;  // Annotation trick to not allow passing Commit objects where CommitMetadata is expected
}

export interface CommitData extends Omit<CommitMetadataData, 'delta_data'> {
    delta_data: DeltaData;
}

export class CommitMetadata {
    id: string;
    parentId: string;
    snapshotHash: string;
    timestamp: number;
    message: string;

    constructor(data: CommitMetadataData) {
        this.id = data.id;
        this.parentId = data.parent_id;
        this.snapshotHash = data.snapshot_hash;
        this.timestamp = data.timestamp;
        this.message = data.message;
    }

    data(): CommitMetadataData {
        return {
            id: this.id,
            parent_id: this.parentId,
            snapshot_hash: this.snapshotHash,
            timestamp: this.timestamp,
            message: this.message,
        };
    }
}

export class Commit {
    id: string;
    parentId: string;
    snapshotHash: string;
    timestamp: number;
    message: string;
    deltaData: DeltaData;

    constructor(data: CommitData) {
        this.id = data.id;
        this.parentId = data.parent_id;
        this.snapshotHash = data.snapshot_hash;
        this.timestamp = data.timestamp;
        this.message = data.message;
        this.deltaData = data.delta_data;
    }

    data(): CommitData {
        return {
            id: this.id,
            parent_id: this.parentId,
            snapshot_hash: this.snapshotHash,
            timestamp: this.timestamp,
            message: this.message,
            delta_data: this.deltaData,
        };
    }

    get delta(): Delta {
        return new Delta(this.deltaData);
    }

    metadata(): CommitMetadata {
        let { delta_data: deltaData, ...metadata } = this.data();
        return new CommitMetadata(metadata);
    }
}

File: py-src/fusion/storage/commit.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from fusion.storage.delta import Delta, DeltaData


@dataclass
class CommitMetadata:
    id: str
    parent_id: str
    snapshot_hash: str
    timestamp: float  # milliseconds since epoch (matches TS Date.now())
    message: str

    def asdict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "parent_id": self.parent_id,
            "snapshot_hash": self.snapshot_hash,
            "timestamp": self.timestamp,
            "message": self.message,
        }

    @staticmethod
    def from_dict(data: dict[str, Any]) -> CommitMetadata:
        return CommitMetadata(
            id=data["id"],
            parent_id=data["parent_id"],
            snapshot_hash=data["snapshot_hash"],
            timestamp=data["timestamp"],
            message=data["message"],
        )


@dataclass
class Commit:
    id: str
    parent_id: str
    snapshot_hash: str
    timestamp: float  # milliseconds since epoch
    message: str
    delta_data: DeltaData = field(default_factory=dict)

    def asdict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "parent_id": self.parent_id,
            "snapshot_hash": self.snapshot_hash,
            "timestamp": self.timestamp,
            "message": self.message,
            "delta_data": self.delta_data,
        }

    @staticmethod
    def from_dict(data: dict[str, Any]) -> Commit:
        return Commit(
            id=data["id"],
            parent_id=data["parent_id"],
            snapshot_hash=data["snapshot_hash"],
            timestamp=data["timestamp"],
            message=data["message"],
            delta_data=data.get("delta_data", {}),
        )

    @property
    def delta(self) -> Delta:
        return Delta.from_data(self.delta_data)

    def metadata(self) -> CommitMetadata:
        return CommitMetadata(
            id=self.id,
            parent_id=self.parent_id,
            snapshot_hash=self.snapshot_hash,
            timestamp=self.timestamp,
            message=self.message,
        )
